### DISCUSSION TITLE
how to export full BCR receptors with mutated nt?

### DISCUSSION THREAD

**Discussion by Angieiscute** [milaboratory/mixcr#1069] [Q&A] [answered]
`mixcr exportClones   -p fullImputed` seems only export the sequences without mutations. Even if there is a mutation point in the vdjca file, which I can see with `mixcr exportAlignmentsPretty`. 

Is this true?

Thanks!

**Comment by mizraelson** [comment, marked as answer]
Hi,
First of all, judging by the command, I think you use an outdated version of MiXCR. I would strongly recommend updating to the latest version as it has lots of improvements.

Regarding your question:

`mixcr exportClones   -p fullImputed` returns clonotype tables with gene features imputed with germline nucleotides if a particular region (or a part of the region) is not covered. Lets say you library only partially covers FR1 region of V gene, then `-p fullImputed` will return the FR1 with capital letters for the part covered with the read and lower case letters imputed from the germline. In other words imputation doesn't have anything to do with hypermutations.

https://mixcr.com/mixcr/reference/mixcr-export/#export-contigs-with-imputation

When you use `mixcr exportAlignmentsPretty`, you look at the alignments of raw reads to the reference sequence. Here some mutations are the result of PCR and sequencing errors and others are true hypermutations. `mixcr assemble` assembles these alignments into clones and corrects the errors (not hypermutations).

If you are looking for a full-length BCR receptor reconstruction you want to specify `--assemble-clonotypes-by VDJRegion` parameter in `mixcr assemble` step. That means that two clones that differ somewhere in the VDJRegion (from FR1 till FR4) will be considered as two different clones. Thus in the output clonotype table you will have two records (with the same CDR3 but some different mutations within the VDJRegion).

I can help you with the command for a new MiXCR. Can you tell more about the cDNA library architecture that you use? Is it a commercial kit or a custom protocol?

**Reply by Angieiscute** [reply]
wow, thank you mizraelson for your detailed reply. I'm using cDNA and 5RACE-UMI library architecture, it's our lab protocol not a commercial kit.

**Reply by mizraelson** [reply]
And where is the UMI located and how long is it?

**Reply by Angieiscute** [reply]
I have to double-check with my lab colleagues, but we use MIGEC for the demultiplexing. The fastq files I got have already been UMI processed. The mixcr commands I got from my labmate are as following,

```
mixcr align --species mmu -OallowPartialAlignments=true -OvParameters.geneFeatureToAlign=VTranscript -OreadsLayout=Collinear R1.fastq.gz R2.fastq.gz alignment.vdjca
mixcr assemble --write-alignments -OcloneClusteringParameters=null alignment.vdjca alignment.clna
mixcr  assembleContigs alignment.clna alignment.clns
mixcr  exportClones --chains IGH --preset fullImputed alignment.clns IGH.txt
```

**Reply by mizraelson** [reply]
I see. MiGEC is our legacy software, currently MiXCR supports UMI processing, so I suggest using it for the whole pipeline.
I also see you use `mixcr assembleContigs` which is required in case of fragmented data (e.g. rna-seq, 10x genomics single cell etc.) but doesn't really do anything for 5'RACE amplicon data.

Practically, the new MiXCR can process all your data starting from raw reads in one command.

If you can share the barcode structure:
The UMI location and length, what read is it in(R1 or R2?)
Are there any adaptor sequences present?
And I assume the reverse primer is in the C gene?

I will provide you with the one single line command for your data.

**Reply by Angieiscute** [reply]
`assembleContigs` is to get the `fullImputed` output data. As I got the information from my lab, our umi is in the 5' adaptor in R2 with a length of 16bp. The reverse primer is in the C region.

**Reply by mizraelson** [reply]
You don't need `assembleContigs` to impute the germline.

Does the R2 start from the UMI or are there any other adaptor sequence present before and/or after the UMI.

**Reply by Angieiscute** [reply]
There are 

> You don't need `assembleContigs` to impute the germline.
Same in the old version of mixcr3?

> Does the R2 start from the UMI or are there any other adaptor sequence present before and/or after the UMI.
There is a 11bp adaptor before and 3bp after the umi.

**Reply by mizraelson** [reply]
JeYes, `assembleСontigs` assembles alignments that initially didn't cover CDR3 but can be still used to recreate the full-length (which is the case with RNA-seq data).

With the new mixcr you can run:

```

mixcr analyze generic-bcr-amplicon-umi \
    --species mmu \
    --rna \
    --tag-pattern "^(R1:*) \ ^N{11}(UMI:N{16})N{3}(R2:*)" \
    --rigid-left-alignment-boundary \
    --floating-right-alignment-boundary C \
    --assemble-clonotypes-by VDJRegion \
    --impute-germline-on-export \
      input_R1.fastq.gz \
      input_R2.fastq.gz \
      result
```

The command above should do the trick. The UMI correction works significantly better now.

Also, in your original command, what was the reason for:

`-OcloneClusteringParameters=null`

**Reply by Angieiscute** [reply]
Thanks! I will download mixcr 4 and give it a try. 

`-OcloneClusteringParameters=null` parameter will switch off the frequency-based correction of PCR errors.

**Reply by mizraelson** [reply]
Why did you want to turn it off? If needed I can show you how to turn it off with this new command, but I don't recommend it in most cases.

**Reply by Angieiscute** [reply]
Yes, please show me the new command. We want to turn it off because we think UMI process has already done the umi-based correction.

**Reply by Angieiscute** [reply]
please also show me the command for exporting full BCR receptors with mutated nt. thanks!

**Reply by mizraelson** [reply]
Ok, 

So to turn clustering off you can use the following parameter:
`-Massemble.cloneAssemblerParameters.cloneClusteringParameters=null`

Again, UMIs and clustering are two complimentary steps not mutually exclusive, thus I would not recommend using it as it most likely will introduce errors in the final results. 

The following command will assemble clonotypes by VDJRegion, thus, if two clones differ by one or several hypermutations, you will see two different records in the clonotype table:

```
mixcr analyze generic-bcr-amplicon-umi \
    --species mmu \
    --rna \
    --tag-pattern "^(R1:*) \ ^N{11}(UMI:N{16})N{3}(R2:*)" \
    --rigid-left-alignment-boundary \
    --floating-right-alignment-boundary C \
    --assemble-clonotypes-by VDJRegion \
    --impute-germline-on-export \
      input_R1.fastq.gz \
      input_R2.fastq.gz \
      result
```

The column `targetSequences` represents assembled full-length VDJRegion of an actual clone. 

In the tsv clonotype table you will see alignments columns:
`allVAlignments 
  allDAlignments 
  allJAlignments
  allCAlignments
`
These columns will include information on mutations relative to a reference gene.  (e.g. 0|290|314|16|306|ST5GSA50CSA69GSC82GSA83CSG134ASA143TSG146T|1338.0)

Reference points in the alignments columns are given relative to the `targetSequences`

In our docs we have all the fields [explained](https://mixcr.com/mixcr/reference/ref-mutations-encoding/).
Let me know if you have other questions.

**Reply by Angieiscute** [reply]
got u! thanks a lot!
